File: cpp/src/0234.cpp
/*
[234. Palindrome Linked List (Easy)](https://leetcode.com/problems/palindrome-linked-list/)

@topic: Linked List
@topic: Two Pointers
@topic: Stack
@topic: Recursion

Constraints:
  - The number of nodes in the list is in the range `[1, 10^5]`.
  - `0 <= Node.val <= 9`
*/

#include <stack>

#include "structs/LinkedList.hpp"

using std::stack;

class Solution {
   public:
    bool isPalindrome(ListNode* head) {
        if (!head->next) return true;
        stack<int> stk;
        auto slow = head;
        auto fast = head;
        while (fast && fast->next) {
            stk.push(slow->val);
            slow = slow->next;
            fast = fast->next->next;
        }
        if (fast) {
            slow = slow->next;
        }
        while (slow) {
            if (stk.top() != slow->val) {
                return false;
            }
            stk.pop();
            slow = slow->next;
        }
        return true;
    }
};

#include <cassert>
#include <chrono>
#include <iostream>
#include <vector>

using std::tuple;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    using namespace LinkedList;

    Solution o;

    vector<tuple<ListNode*, bool>> CASES = {
        {of({1, 2, 2, 1}), true},
        {of({1, 2}), false},
    };

    for (auto& [head, excepted] : CASES) {
        assert(o.isPalindrome(head) == excepted);
    }

    auto start = system_clock::now();
    for (auto& [head, _] : CASES) {
        for (auto i = 0; i < 100000; ++i) {
            o.isPalindrome(head);
        }
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}

File: python/src/0234.py
from collections import deque
from typing import Optional

from structs.ListNode import ListNode, create


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        stack = deque()
        fast = slow = head
        while fast and fast.next:
            stack.append(slow.val)
            slow = slow.next
            fast = fast.next.next
        if fast:
            slow = slow.next
        while slow:
            if stack.pop() != slow.val:
                return False
            slow = slow.next
        return True

    # def isPalindrome(self, head: ListNode) -> bool:
    #     if not head or not head.next:
    #         return True
    #     fast = slow = head
    #     while fast and fast.next:
    #         slow = slow.next
    #         fast = fast.next.next
    #     left = head
    #     right = slow = self.reverse(slow.next if fast else slow)
    #     while right and left.val == right.val:
    #         left = left.next
    #         right = right.next
    #     self.reverse(slow)
    #     return right is None

    # def reverse(self, head: Optional[ListNode]) -> Optional[ListNode]:
    #     prev = None
    #     walk = head
    #     while walk:
    #         next = walk.next
    #         walk.next = prev
    #         prev = walk
    #         walk = next
    #     return prev


if __name__ == '__main__':
    import unittest

    CASES = (
        (create(1, 2, 3, 2, 1), True),
        (create(1, 2, 2, 1), True),
        (create(1, 2), False),
        (create(1), True),
        (None, True),
    )

    class Test(unittest.TestCase):
        def test(self):
            isPalindrome = Solution().isPalindrome
            for head, excepted in CASES:
                actual = isPalindrome(head)
                self.assertEqual(actual, excepted)

    unittest.main()
